### PR TITLE
Review (DO NOT MERGE)

### PR DESCRIPTION
**Pull request by lundibundi**
/cc @tshemsedinov

### CODE REVIEW

File: matrixClass.js [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+  }
+
+  static isValid(matrix) {
+    if (!Array.isArray(matrix)) {
+      return false;
+    }
+    if (matrix.length < 1) {
+      return false;
+    }
+    const mainLen = matrix[0].length;
```
Comment (lundibundi) [inline matrixClass.js:20]: This never checked that `matrix[i]` is an Array. Would most likely behave incorrectly (return true) on something like `[1, 2, 3]`.

File: matrixClass.js [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+      }
+    }
+    return new Matrix(res);
+  }
+
+  static areEqual(matr1, matr2) {
+    if (!Matrix.haveOneSize(matr1, matr2)) return false;
+    for (let i = 0; i < matr1.matrix.length; i++) {
+      for (let j = 0; j < matr1.matrix[0].length; j++) {
+        if (!(matr1.matrix[i][j] === matr2.matrix[i][j])) return false;
```
Comment (lundibundi) [inline matrixClass.js]:
```suggestion
        if (matr1.matrix[i][j] !== matr2.matrix[i][j]) return false;
```
?

File: matrixClass.js [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+    let res = 'Solution:\n';
+    for (let i = 0; i < step2Matr.matrix.length; i++) {
+      const coeff = step2Matr.matrix[i][i];
+      const sum = step2Matr.matrix[i][columnNum - 1];
+      const value = RatFract.div(sum, coeff);
+      res += `  X${i + 1} = ${value}\n`;
+    }
+    return res;
+  }
+}
```
Comment (lundibundi) [inline matrixClass.js]: These should better be outside of Matrix class since they are not directly related to matrix.

File: matrixClass.js [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+  static findSolOfSoLE(extendMatr) {
+    const columnIndex = extendMatr.matrix[0].length - 2;
+    const systHasntSol = Matrix.systemHasntSol(extendMatr);
+    if (systHasntSol) {
+      return systHasntSol;
+    }
+    const step1Matr = extendMatr.getStepMatr(columnIndex);
+    const step2Matr = step1Matr.getStepMatrReversed(columnIndex);
+    const columnNum = step2Matr.matrix[0].length;
+    let res = 'Solution:\n';
```
Comment (lundibundi) [inline matrixClass.js]: The solver should not concern itself with the way results are presented, this should just return the results and not a string.

File: simpleFunctions.js [hunk 5]
```diff
@@ -0,0 +1,230 @@
+'use strict'
+
+const ratFract = require('./rationalFractions.js')
```
Comment (lundibundi) [inline simpleFunctions.js]: This file looks unrelated to the other ones? (it just defines some matrix manipulation functions not used anywhere else)

File: index.js [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+
+const addMatrix = (msg, match) => {
+  const chatId = msg.chat.id;
+  const name = match[1];
+  bot.sendMessage(chatId, `${answ.sendName}${name}`);
+  bot.sendMessage(chatId, answ.enterMatr);
+  bot.once('message', msg => inputMatrix(msg, name));
+};
+
+bot.onText(/\/start/, msg => {
```
Comment (lundibundi) [inline index.js]: This could use some better code separation, i.e. have commands folder where each file would describe 1 command and export appropriate function(s)/name and then gather and register them here.

File: index.js [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+    const matr = matrices[chatId][name];
+    const transposedMatr = matr.transpose();
+    const text = matrixToText(transposedMatr.matrix);
+    bot.sendMessage(chatId, text);
+  }
+};
+
+const invertMatr = (msg, match) => {
+  const chatId = msg.chat.id;
+  const name = match[1];
```
Comment (lundibundi) [inline index.js]: Each of the functions performs the same 'argument-extraction', it would be better to at least encapsulate it in some function. Or better have a wrapper function that would separate and extract arguments from a user query and call these functions with just arguments.
i.e.
```js
const commandWrapper = (handler, msg, match) => {
  const chatId = msg.chat.id;
  const result = handler(chatId, ...match.slice(1));
  bot.sendMessage(chatId, result);
};
```

**Review by lundibundi — COMMENTED** [review]
Looks good, could use better code abstraction and separation.

File: matrixClass.js (hunk 1 above)
Comment (vladyslavry141) [reply matrixClass.js:20]: [Done](https://github.com/vladyslavry141/matrix-bot/commit/5311e57b310fe39b02f09f8a2e1e00205bb608d1)

File: matrixClass.js (hunk 3 above)
Comment (vladyslavry141) [reply matrixClass.js]: [Create separate class for equations](https://github.com/vladyslavry141/matrix-bot/commit/3a04fdb29b92aa4d41df57b0c2c4073f844d6bd1)

File: index.js (hunk 6 above)
Comment (vladyslavry141) [reply index.js]: [Create one object with commands names and functions](https://github.com/vladyslavry141/matrix-bot/commit/5dfe3f5d189a68d25ba2ff6859edc336e7be37b8)

File: index.js (hunk 7 above)
Comment (vladyslavry141) [reply index.js]: [Add one massage listener which extract chatId and call function with it ](https://github.com/vladyslavry141/matrix-bot/commit/5dfe3f5d189a68d25ba2ff6859edc336e7be37b8)

File: simpleFunctions.js (hunk 5 above)
Comment (vladyslavry141) [reply simpleFunctions.js]: It was a test file which i forgot to delete
